tools/CMakeLists.txt

```diff
@@ -32,16 +32,27 @@ set(IREE_PACKAGE_ROOT_PREFIX "iree/tools")
 iree_add_all_subdirs()
 set(IREE_PACKAGE_ROOT_PREFIX "")
 
-# TODO(scotttodd): Should this be checking IREE_BUILD_COMPILER?
-#   Maybe we should disallow setting both at the same time, since it's
-#   ambigious which should be used
+# If cross-compiling and not building the compiler, then attempt to find
+# the compiler tools.
+# This is actual broken because the situation is tri-state:
+#   1. Cross-compiling with no built compiler: Should work the same as
+#      IREE_BUILD_COMPILER=OFF in a host build (i.e. nothing depending
+#      on the compiler should be built).
+#   2. Cross-compiling with a compiler built for the target: Anything
+#      on the host which needs the compiler, still must have host tools.
+#   3. Normal host build.
+# This simplistic setup makes #2 impossible and it overloads #1 to
+# also support building things that depend on the compiler. The targets
+# need to be aliased/forked for host variants to fully support. For now
+# we just make all of these as OPTIONAL and let things break if not set up
+# right.
 if(IREE_HOST_BIN_DIR AND NOT IREE_BUILD_COMPILER)
-  iree_import_binary(NAME iree-tblgen)
-  iree_import_binary(NAME iree-compile)
-  iree_import_binary(NAME iree-opt)
-  iree_import_binary(NAME iree-run-mlir)
-  iree_import_binary(NAME clang)
-  iree_import_binary(NAME llvm-link)
+  iree_import_binary(NAME iree-tblgen OPTIONAL)
+  iree_import_binary(NAME iree-compile OPTIONAL)
```

Comment:
Thanks, adding `OPTIONAL` to `iree_import_binary` here was how I was going to solve this too.

> and let things break if not set up right

The idea with https://github.com/openxla/iree/pull/16086 is that the "let things break" case can be a soft failure, with tests appearing as "disabled" when running `ctest` if parts of the configuration are missing. I prefer that over a silent failure (tests not being added at all) or hard failure (configure error). We should keep cases needing compiler tools limited to "e2e" tests and samples so these sorts of lines are easier to draw in the build system.

We could have `IREE_DEV_MODE` or some other option change soft failures to hard failures for CI / developers if we want to enforce that enough of the configuration be explicit. I think with so many possible combinations of settings just leaving enough breadcrumbs in configure, build, and test output should be enough though.

Reply:
Yeah, let's chat when we have some attention to spare... I think I see what needs to be done too, but not completely trivial.